File: mobile/modules/expo-waveform-extractor/android/src/main/java/expo/modules/waveformextractor/ExpoWaveformExtractorModule.kt
package expo.modules.waveformextractor

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoWaveformExtractorModule : Module() {
  private val extractors = mutableMapOf<String, WaveformExtractor>()

  override fun definition() = ModuleDefinition {
    Name("ExpoWaveformExtractorModule")

    Events("onCurrentExtractedWaveformData")

    AsyncFunction("extractWaveform") { audioUri: String, samplesPerPixel: Int, playerKey: String, promise: Promise ->
      try {
        android.util.Log.d("WaveformExtractor", "🎵 Starting extraction: $audioUri, samples: $samplesPerPixel")

        val extractor = WaveformExtractor(
          path = audioUri,
          expectedPoints = samplesPerPixel,
          key = playerKey,
          extractorCallBack = object : ExtractorCallBack {
            override fun onProgress(value: Float) {
              android.util.Log.d("WaveformExtractor", "📊 Progress: ${(value * 100).toInt()}%")
            }

            override fun onReject(error: String?, message: String?) {
              android.util.Log.e("WaveformExtractor", "❌ Error: $error - $message")
              promise.reject("EXTRACTION_ERROR", "$error: $message", null)
              extractors.remove(playerKey)
            }

            override fun onResolve(value: MutableList<MutableList<Float>>) {
              android.util.Log.d("WaveformExtractor", "✅ Extraction complete! Samples: ${if (value.isNotEmpty()) value[0].size else 0}")
              val result = if (value.isNotEmpty()) value[0] else emptyList<Float>()
              promise.resolve(result)
              extractors.remove(playerKey)
            }

            override fun onForceStop() {
              android.util.Log.d("WaveformExtractor", "⏹️ Extraction stopped")
              extractors.remove(playerKey)
            }
          }
        )

        extractors[playerKey] = extractor
        extractor.startDecode()
        android.util.Log.d("WaveformExtractor", "🚀 Decoder started")
      } catch (e: Exception) {
        android.util.Log.e("WaveformExtractor", "💥 Init error: ${e.message}")
        promise.reject("INIT_ERROR", "Failed to initialize extractor: ${e.message}", e)
      }
    }

    AsyncFunction("cancelExtraction") { playerKey: String ->
      extractors[playerKey]?.forceStop()
      extractors.remove(playerKey)
    }

    OnDestroy {
      extractors.values.forEach { it.forceStop() }
      extractors.clear()
    }
  }
}

File: mobile/modules/expo-waveform-extractor/ios/ExpoWaveformExtractorModule.swift
import ExpoModulesCore

public class ExpoWaveformExtractorModule: Module {
  private var extractors: [String: WaveformExtractor] = [:]

  public func definition() -> ModuleDefinition {
    Name("ExpoWaveformExtractorModule")

    Events("onCurrentExtractedWaveformData")

    AsyncFunction("extractWaveform") { (audioUri: String, samplesPerPixel: Int, playerKey: String, promise: Promise) in
      guard let url = URL(string: audioUri) else {
        promise.reject("INVALID_URI", "Invalid audio URI: \(audioUri)")
        return
      }

      // Convert file:// URI to path
      let fileURL: URL
      if url.scheme == "file" {
        fileURL = url
      } else {
        fileURL = URL(fileURLWithPath: audioUri)
      }

      do {
        let extractor = try WaveformExtractor(
          url: fileURL,
          module: self,
          playerKey: playerKey,
          resolve: promise.resolve,
          reject: promise.reject
        )

        self.extractors[playerKey] = extractor

        // Extract waveform asynchronously
        DispatchQueue.global(qos: .userInitiated).async {
          let data = extractor.extractWaveform(
            samplesPerPixel: samplesPerPixel,
            playerKey: playerKey
          )

          if let waveformData = data {
            // Get channel mean (mono or stereo averaged)
            let meanData = extractor.getChannelMean(data: waveformData)
            promise.resolve(meanData)
          } else {
            promise.reject("EXTRACTION_FAILED", "Failed to extract waveform data")
          }

          self.extractors.removeValue(forKey: playerKey)
        }
      } catch {
        promise.reject("INIT_ERROR", "Failed to initialize extractor: \(error.localizedDescription)")
      }
    }

    AsyncFunction("cancelExtraction") { (playerKey: String) in
      self.extractors[playerKey]?.cancel()
      self.extractors.removeValue(forKey: playerKey)
    }

    OnDestroy {
      self.extractors.values.forEach { $0.cancel() }
      self.extractors.removeAll()
    }
  }

  func sendEvent(name: String, body: Any?) {
    sendEvent(name, body)
  }
}
